build_tools/llvm/mlir_config.cmake

```diff
@@ -0,0 +1,58 @@
+# export MLIR_BUILD_DIR=$HOME/tmp/mlirbuild
+# export MLIR_INSTALL_DIR=$HOME/tmp/mlirroot
+# cmake -G Ninja \
+#     -B "$MLIR_BUILD_DIR" -S third_party/llvm-project/mlir \
+#     -DLLVM_DIR="${LLVM_INSTALL_DIR}/lib/cmake/llvm" \
+#     -C build_tools/llvm/mlir_config.cmake \
+#     -DCMAKE_BUILD_TYPE="Release" \
+#     -DPython3_EXECUTABLE='$(which $python3_command)' \
+#     -DMLIR_ENABLE_BINDINGS_PYTHON=ON \
+#     -DCMAKE_INSTALL_PREFIX="$MLIR_INSTALL_DIR" \
+#     -DCMAKE_C_COMPILER_LAUNCHER=ccache \
+#     -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
+#     -DCMAKE_C_COMPILER=clang \
+#     -DCMAKE_CXX_COMPILER=clang++ \
+#     -DCMAKE_LINKER=lld
+# TODO: For some reason Python sources aren't selectable as part of a
+# distribution and targets don't show up properly.
```

Comment:
`mlir-<distribution>-cmake-exports` needs to be included in `LLVM_MlirDevelopment_DISTRIBUTION_COMPONENTS` (which for this distribution would be `mlir-mlirdevelopment-cmake-exports`).

This will distribute the `lib/cmake/mlir/MLIRMlirDevelopmentTargets.cmake` file which contains all the imported targets. `lib/cmake/mlir/MLIRConfig.cmake` will include this file, so that any project which has a `find_package` dependency on the installed MLIR can use these targets.

```cmake
# From lib/cmake/mlir/MLIRConfig.cmake
...
# Provide all our library targets to users.
# More specifically, configure MLIR so that it can be directly included in a top
# level CMakeLists.txt, but also so that it can be imported via `find_package`.
# This is based on how LLVM handles exports.
if(NOT TARGET MLIRSupport)
  include("${MLIR_CMAKE_DIR}/MLIRMlirDevelopmentTargets.cmake" OPTIONAL)
endif()
```

Reply:
For example, here is how the `MLIRSupport` target is imported by `lib/cmake/mlir/MLIRMlirDevelopmentTargets.cmake`.

```cmake
# From lib/cmake/mlir/MLIRMlirDevelopmentTargets.cmake
...
# Create imported target MLIRSupport
add_library(MLIRSupport STATIC IMPORTED)

set_target_properties(MLIRSupport PROPERTIES
  INTERFACE_LINK_LIBRARIES "LLVM"
)
...
```

Reply:
Relevant section from [Building a Distribution of LLVM](https://llvm.org/docs/BuildingADistribution.html#multi-distribution-configurations)

> Each distribution creates its own set of CMake exports, and the target to install the CMake exports for a particular distribution for a project is named ${project}-${distribution}-cmake-exports, where ${project} is the name of the project in lowercase and ${distribution} is the name of the distribution in lowercase, unless the project is LLVM, in which case the target is just named ${distribution}-cmake-exports. These targets need to be explicitly included in the LLVM_<distribution>_DISTRIBUTION_COMPONENTS variable in order to be included as part of the distribution.

Reply:
Good catch. I thought I actually had that, which is why I was scratching my head about why it wasn't working :)

Reply:
I think I'm going to land this as-is and then adjust in a followup. Would like to fiddle with that a little bit with the CI apart from this change.